lib/using_restricted_array.rb

```diff
@@ -1,61 +1,118 @@
 require_relative 'restricted_array.rb'
+require 'pry'
 # RestrictedArray can be created using a specified size, or a random size in
 # the range of 1-20 will be chosen for you.
 # All values are integers in the range of 1-221.
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array or the length of the input array
+# Space complexity:
 def length(array)
-  raise NotImplementedError
+  #delete NotImplementedError and write your code there instead
+  i = 0
+
+  until array[i] == nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
+# Time complexity: quadratic, or O(n^2), where n is the number of items in the array or the length of the input array
 # Space complexity: ?
 def print_array(array)
-  raise NotImplementedError
+  len = length(array)
+
+  i = 0
+  while i < len
+    puts array[i]
+    print " "
+    i += 1
+  end
+
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
+# Time complexity: O(n), where n is the number of values in the array
 # Space complexity: ?
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    return true if array[i] == value_to_find
+  end
+
+  return false
+
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
+# Time complexity: O(n)
 # Space complexity: ?
 def find_largest(array, length)
-  raise NotImplementedError
+  largest_number = array[0]
+  length.times do |i|
+    if array[i] > largest_number
+      largest_number =  array[i]
+    end
+  end
+  return largest_number
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
+# Time complexity: O(n)
 # Space complexity: ?
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest_number = array[0]
+  length.times do |i|
+    if array[i] < smallest_number
+      smallest_number = array[i]
+    end
+  end
+  return smallest_number
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
+# Time complexity: O(n)
 # Space complexity: ?
 def reverse(array, length)
-  raise NotImplementedError
+  i = 0
+
+  for i in 0...length/2
+    temp = array[i]
+    array[i] = array[length - i -1]
+    array[length -i -1] = temp
+  end
 end
+ 
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
+# Time complexity: O(log n)? O(1)?
```

Comment:
`O(lg n)` is correct. 🙂

(In general, you'll see `lg` when you're repeatedly multiplying the work by some fraction - e.g. `1/2`, `3/4`, etc. Binary search cuts its workload in half (= `1/2` = `0.5`) with each step - hence the `O(lg n)`. 🙂)